tensorflow/core/common_runtime/bfc_allocator.cc

```diff
@@ -260,6 +261,76 @@ size_t BFCAllocator::RoundedBytes(size_t bytes) {
   return rounded_bytes;
 }
 
+bool BFCAllocator::DeallocateFreeRegions(size_t rounded_bytes) {
+  // Searching for free regions.
+  absl::flat_hash_set<void*> free_region_ptrs;
+  size_t total_free_bytes = 0;
+  for (const auto& region : region_manager_.regions()) {
+    ChunkHandle h = region_manager_.get_handle(region.ptr());
+    bool any_use = false;
+    while (h != kInvalidChunkHandle) {
+      const Chunk* c = ChunkFromHandle(h);
+      if (c->in_use()) {
+        any_use = true;
+        break;
+      }
+      h = c->next;
+    }
+
+    if (!any_use) {
+      VLOG(2) << "Found free region with ptr = " << region.ptr();
+      free_region_ptrs.insert(region.ptr());
+      total_free_bytes += region.memory_size();
+    }
+  }
+
+  if (total_free_bytes == 0) {
+    return false;
+  }
+
+  // Rough estimation to check whether deallocation can help.
+  size_t available_bytes =
+      memory_limit_ - total_region_allocated_bytes_ + total_free_bytes;
+  if (rounded_bytes > available_bytes) {
+    return false;
+  }
+
+  VLOG(INFO) << "Re-allocate memory regions to avoid OOM due to memory"
+             << " fragmentation. If you see this message frequently, note"
+             << " that the re-allocation may incur performance overhead despite"
+             << " better memory utilization. You may try smaller batch sizes"
+             << " to see if it can give you better performance.";
+
+  // Deallocate free regions.
+  auto it = region_manager_.regions().begin();
+  while (it != region_manager_.regions().end()) {
```

Comment:
Nitpick: IMO it would be more readable to have this loop in a static helper function.